source/distributed/repartitioning_policy_tools.cc

```diff
@@ -0,0 +1,231 @@
+                is_coarse,
+                communicator,
+                owning_ranks_of_coarse_cells,
+                false);
+
+      Utilities::MPI::ConsensusAlgorithms::Selector<
+        std::pair<types::global_cell_index, types::global_cell_index>,
+        unsigned int>
+        consensus_algorithm(process, communicator);
+      consensus_algorithm.run();
+    }
+
+    const auto tria =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &tria_coarse_in);
+
+    Assert(tria, ExcNotImplemented());
+
+    LinearAlgebra::distributed::Vector<double> partition(
+      tria->global_active_cell_index_partitioner().lock());
+
+    for (const auto &cell : tria_coarse_in.active_cell_iterators())
+      if (cell->is_locally_owned())
+        partition[cell->global_active_cell_index()] =
+          owning_ranks_of_coarse_cells[is_coarse.index_within_set(
+            cell_id_translator.translate(cell))];
+
+    partition.update_ghost_values();
+
+    return partition;
+  }
+
+
+  template <int dim, int spacedim>
+  MinimalGranularityPolicy<dim, spacedim>::MinimalGranularityPolicy(
+    const unsigned int n_min_cells)
+    : n_min_cells(n_min_cells)
+  {}
+
+
+
+  template <int dim, int spacedim>
+  LinearAlgebra::distributed::Vector<double>
+  MinimalGranularityPolicy<dim, spacedim>::partition(
+    const Triangulation<dim, spacedim> &tria_in) const
+  {
+    const auto tria =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &tria_in);
+
+    Assert(tria, ExcNotImplemented());
+
+    LinearAlgebra::distributed::Vector<double> partition(
+      tria->global_active_cell_index_partitioner().lock());
+
+    // step 1) check if all processes have enough cells
+
+    unsigned int n_locally_owned_active_cells = 0;
+    for (const auto &cell : tria_in.active_cell_iterators())
+      if (cell->is_locally_owned())
+        ++n_locally_owned_active_cells;
+
+    const auto comm = tria_in.get_communicator();
+
+    if (Utilities::MPI::min(n_locally_owned_active_cells, comm) >= n_min_cells)
+      return {}; // all processes have enough cells
+
+    // step 2) there are processes which do not have enough processes so that
+    // a repartitioning kicks in with the aim that all processes that own
+    // cells have at least the specified number of cells
+
+    const unsigned int n_global_active_cells = tria_in.n_global_active_cells();
+
+    const unsigned int n_partitions =
+      std::max<unsigned int>(1,
+                             std::min(n_global_active_cells / n_min_cells,
+                                      Utilities::MPI::n_mpi_processes(comm)));
+
+    const unsigned int min_cells = n_global_active_cells / n_partitions;
+
+    const auto convert = [&](const unsigned int i) {
+      if (i < (i / min_cells) * min_cells + (n_global_active_cells % min_cells))
+        return i / (min_cells + 1);
+      else
+        return (i - (n_global_active_cells % min_cells)) / min_cells;
+    };
```

Comment:
Nice; maybe add a short comment where this formula comes from to help the reader.